components/server/src/authorization/relationship-updater.ts

```diff
@@ -0,0 +1,90 @@
+/**
+ * Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { ProjectDB, TeamDB, UserDB } from "@gitpod/gitpod-db/lib";
+import { AdditionalUserData, Organization, User } from "@gitpod/gitpod-protocol";
+import { log } from "@gitpod/gitpod-protocol/lib/util/logging";
+import { inject, injectable } from "inversify";
+import { RedisMutex } from "../redis/mutex";
+import { Authorizer } from "./authorizer";
+import { ApplicationError, ErrorCodes } from "@gitpod/gitpod-protocol/lib/messaging/error";
+
+@injectable()
+export class RelationshipUpdater {
+    public readonly version = 1;
+
+    constructor(
+        @inject(RedisMutex) private readonly mutex: RedisMutex,
+        @inject(UserDB) private readonly userDB: UserDB,
+        @inject(TeamDB) private readonly orgDB: TeamDB,
+        @inject(ProjectDB) private readonly projectDB: ProjectDB,
+        @inject(Authorizer) private readonly authorizer: Authorizer,
+    ) {}
+
+    /**
+     * Updates all relationships for a user according to the current state of the database.
+     * @param user
+     * @returns
+     */
+    public async migrate(user: User): Promise<User> {
+        if (user.additionalData?.fgaRelationshipsVersion === this.version) {
+            return user;
+        }
+        return this.mutex.using([`fga-migration-${user.id}`], 2000, async () => {
+            const before = new Date().getTime();
+
+            const updatedUser = await this.userDB.findUserById(user.id);
+            if (!updatedUser) {
+                throw new ApplicationError(ErrorCodes.NOT_FOUND, "User not found");
+            }
+            user = updatedUser;
```

Comment:
Just pondering about `deleted` users... but I think it's good to migrate them as well :+1:

Reply:
I think we should not handle any deleted entities but remove all relationships on delete. Let's handle that in follow up changes.